### PR TITLE
assets: add file system layer for zipped embed assets

### PR DESCRIPTION
**Pull request by jan--f**
This commits adds an optional decompression layer to a embed.FS
instance. If a file can't be found during Open, the filename plus a .gz
ending is tried. If this succeeds, the content is decompressed and
served.

Signed-off-by: Jan Fajerski <jfajersk@redhat.com>

This was originally proposed in https://github.com/prometheus/prometheus/pull/9380.

### CODE REVIEW

**Comment by jan--f** [comment]
I'm not sure how I can keep building and testing this for go < 1.16. Any input welcome.

**Comment by roidelapluie** [comment]
I'd create a separate module like we do for sigv4

**Comment by jan--f** [comment]
Thanks, I didn't think of that.
I also reworked the circleci workflow a little bit. I was trying to avoid a conditional like `if go_version >= 1.16` in the test job, though it ended up a bit more verbose. Alternatively we could add a new parameter like `test_assets` or so.

**Comment by jan--f** [comment]
@roidelapluie mind taking another look at this?

File: .circleci/config.yml [hunk 1]
```diff
@@ -28,10 +25,57 @@ jobs:
     - run: make test
     - run: cd sigv4 && make test
```
Comment (roidelapluie) [inline .circleci/config.yml:26]: cd assets && make test

**Comment by roidelapluie** [comment]
@juliusv did you review this before merging into prometheus/prometheus?

**Comment by juliusv** [comment]
@roidelapluie Ah I did not.

File: README.md [hunk 2]
```diff
@@ -5,6 +5,7 @@ This repository contains Go libraries that are shared across Prometheus
 components and libraries. They are considered internal to Prometheus, without
 any stability guarantees for external usage.
 
+* **assets**: Embeding of static assets with gzip support
```
Comment (roidelapluie) [inline README.md:8]:
```suggestion
* **assets**: Embedding of static assets with gzip support
```

File: assets/embed_gzip.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	c, err := io.ReadAll(gr)
+	if err != nil {
+		return f, err
+	}
+	// and wrap everything in our custom File
+	return &File{file: f, content: c}, nil
+}
+
+type File struct {
+	// the underlying file
```
Comment (roidelapluie) [inline assets/embed_gzip.go:64]:
```suggestion
	// The underlying file.
```

File: assets/embed_gzip.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	file fs.File
+	// the decrompressed content, needed to return an accurate size
```
Comment (roidelapluie) [inline assets/embed_gzip.go:66]:
```suggestion
	// The decrompressed content, needed to return an accurate size.
```

Comment (roidelapluie) [reply assets/embed_gzip.go:66]: It is not needed for this, it is used to read it.

File: assets/embed_gzip.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+func (f File) Stat() (fs.FileInfo, error) {
+	stat, err := f.file.Stat()
+	if err != nil {
+		return stat, err
+	}
+	return FileInfo{stat, int64(len(f.content))}, nil
+}
+
+func (f *File) Read(buf []byte) (n int, err error) {
```
Comment (roidelapluie) [inline assets/embed_gzip.go:80]: can we avoid the named return here? thanks

File: assets/embed_gzip.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+import (
+	"compress/gzip"
+	"embed"
+	"io"
+	"io/fs"
+	"time"
+)
+
+const (
+	zipSuffix = ".gz"
```
Comment (roidelapluie) [inline assets/embed_gzip.go:25]:
```suggestion
	gzipSuffix = ".gz"
```

File: assets/embed_gzip.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+	if err != nil {
+		return f, err
+	}
+	defer gr.Close()
+
+	c, err := io.ReadAll(gr)
+	if err != nil {
+		return f, err
+	}
+	// and wrap everything in our custom File
```
Comment (roidelapluie) [inline assets/embed_gzip.go:59]:
```suggestion
	// Wrap everything in our custom File.
```

File: assets/embed_gzip.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+	var f fs.File
+	if f, err := compressed.embed.Open(path); err == nil {
+		return f, nil
+	}
+	// try opening path with .gz suffix
+	f, err := compressed.embed.Open(path + zipSuffix)
+	if err != nil {
+		return f, err
+	}
+	// read the (decompressed) content into a buffer
```
Comment (roidelapluie) [inline assets/embed_gzip.go:48]:
```suggestion
	// Read the decompressed content into a buffer.
```

File: assets/embed_gzip.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+type FileSystem struct {
+	embed embed.FS
+}
+
+func New(fs embed.FS) FileSystem {
+	return FileSystem{fs}
+}
+
+func (compressed FileSystem) Open(path string) (fs.File, error) {
+	// if we have the file in our embed FS, just return that (could be a
```
Comment (roidelapluie) [inline assets/embed_gzip.go:37]:
```suggestion
	// If we have the file in our embed FS, just return that (could be a
```

File: assets/embed_gzip.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+	// dir)
```
Comment (roidelapluie) [inline assets/embed_gzip.go:38]:
```suggestion
	// dir.)
```
let's have that comment in one line.

File: assets/embed_gzip.go [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+	var f fs.File
+	if f, err := compressed.embed.Open(path); err == nil {
+		return f, nil
+	}
+	// try opening path with .gz suffix
```
Comment (roidelapluie) [inline assets/embed_gzip.go:43]: This comment is not really needed

**Review by roidelapluie — COMMENTED** [review]
Thanks.

It would be nice if exported function would have the interfaces they implement in comments.

**Comment by jan--f** [comment]
Thank you @roidelapluie! PTAL.

**Comment by roidelapluie** [comment]
thanks!

**Pull request merged**